plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java

```diff
@@ -5588,6 +5588,60 @@ private void testSchemaMismatchesWithDereferenceProjections(Session session, Hiv
         }
     }
 
+    @Test
+    public void testReadWithPartitionSchemaMismatch()
+    {
+        testWithAllStorageFormats(this::testReadWithPartitionSchemaMismatch);
+    }
+
+    private void testReadWithPartitionSchemaMismatch(Session session, HiveStorageFormat format)
+    {
+        String tableName = "read_with_partition_schema_mismatch_" + randomNameSuffix();
+        try {
+            if (isMappingByName(format)) {
+                assertUpdate(session, "CREATE TABLE " + tableName + " (a bigint, b bigint, _part bigint) with (format = '" + format + "', partitioned_by=array['_part'])");
+                assertUpdate(session, "INSERT INTO " + tableName + " values (1, 2, 0)", 1); // a, b, _part
+                assertUpdate(session, "ALTER TABLE " + tableName + " DROP COLUMN a");
+                assertUpdate(session, "INSERT INTO " + tableName + " values (11, 10)", 1); // b, _part
+                assertUpdate(session, "ALTER TABLE " + tableName + " ADD COLUMN c bigint");
+                assertUpdate(session, "INSERT INTO " + tableName + " values (21, 22, 20)", 1); // b, c, _part
+                assertQuery(session, "SELECT b, c, _part FROM " + tableName, "VALUES (2, null, 0), (11, null, 10), (21, 22, 20)"); // read all columns
+                assertQuery(session, "SELECT c, _part FROM " + tableName, "VALUES (null, 0), (null, 10), (22, 20)"); // read some columns
+            }
+            else {
+                // we are not dropping columns for format which use index based mapping as logic is confusing and not consistent between different formats
+                assertUpdate(session, "CREATE TABLE " + tableName + " (a bigint, b bigint, c bigint, _part bigint) with (format = '" + format + "', partitioned_by=array['_part'])");
+                assertUpdate(session, "INSERT INTO " + tableName + " values (1, 2, 3, 0)", 1); // a, b, c, _part
+                assertQuery(session, "SELECT a, b, c, _part FROM " + tableName, "VALUES (1, 2, 3, 0)");
+                assertQuery(session, "SELECT a, c, _part FROM " + tableName, "VALUES (1, 3, 0)");
+                assertUpdate(session, "ALTER TABLE " + tableName + " ADD COLUMN d bigint");
+                assertUpdate(session, "INSERT INTO " + tableName + " values (10, 11, 12, 13, 10)", 1); // a, b, c, _part
+                assertQuery(session, "SELECT a, b, c, d, _part FROM " + tableName, "VALUES (1, 2, 3, null, 0), (10, 11, 12, 13, 10)");
+                assertQuery(session, "SELECT a, d, _part FROM " + tableName, "VALUES (1, null, 0), (10, 13, 10)");
+            }
+        }
+        finally {
+            assertUpdate(session, "DROP TABLE IF EXISTS " + tableName);
+        }
+    }
+
+    private boolean isMappingByName(HiveStorageFormat format)
+    {
+        return switch(format) {
+            case PARQUET -> true;
+            case AVRO -> true;
+            case JSON -> true;
+            case ORC -> false;
```

Comment:
`orc_use_column_names` exists

Reply:
Yeah - for ORC we can force using name based mapping and for PARQET index based. I will test that too.